haystack/components/readers/extractive.py

```diff
@@ -206,62 +206,50 @@ def _postprocess(
         mask = torch.triu(mask)  # End shouldn't be before start
         masked_logits = torch.where(mask, logits, -torch.inf)
         probabilities = torch.sigmoid(masked_logits * self.calibration_factor)
 
         flat_probabilities = probabilities.flatten(-2, -1)  # necessary for topk
+
+        # topk can return invalid candidates as well if answers_per_seq > num_valid_candidates
+        # We only keep probability > 0 candidates later on
         candidates = torch.topk(flat_probabilities, answers_per_seq)
         seq_length = logits.shape[-1]
         start_candidates = candidates.indices // seq_length  # Recover indices from flattening
         end_candidates = candidates.indices % seq_length
+        candidates_values = candidates.values.cpu()
         start_candidates = start_candidates.cpu()
         end_candidates = end_candidates.cpu()
 
-        start_candidates_tokens_to_chars = [
-            [encoding.token_to_chars(start) for start in candidates]
-            for candidates, encoding in zip(start_candidates, encodings)
-        ]
-        if missing_start_tokens := [
-            (batch, index)
-            for batch, token_to_chars in enumerate(start_candidates_tokens_to_chars)
-            for index, pair in enumerate(token_to_chars)
-            if pair is None
-        ]:
-            logger.warning("Some start tokens could not be found in the context: %s", missing_start_tokens)
-        start_candidates_char_indices = [
-            [token_to_chars[0] if token_to_chars else None for token_to_chars in candidates]
-            for candidates in start_candidates_tokens_to_chars
-        ]
-
-        end_candidates_tokens_to_chars = [
-            [encoding.token_to_chars(end) for end in candidates]
-            for candidates, encoding in zip(end_candidates, encodings)
-        ]
-        if missing_end_tokens := [
-            (batch, index)
-            for batch, token_to_chars in enumerate(end_candidates_tokens_to_chars)
-            for index, pair in enumerate(token_to_chars)
-            if pair is None
-        ]:
-            logger.warning("Some end tokens could not be found in the context: %s", missing_end_tokens)
-        end_candidates_char_indices = [
-            [token_to_chars[1] if token_to_chars else None for token_to_chars in candidates]
-            for candidates in end_candidates_tokens_to_chars
-        ]
-
-        probabilities = candidates.values.cpu()
-
-        return start_candidates_char_indices, end_candidates_char_indices, probabilities
+        start_candidates_tokens_to_chars = []
+        end_candidates_tokens_to_chars = []
+        for i, (s_candidates, e_candidates, encoding) in enumerate(zip(start_candidates, end_candidates, encodings)):
+            # Those with probabilities > 0 are valid
+            valid = candidates_values[i] > 0
+            s_char_spans = []
+            e_char_spans = []
+            for start, end in zip(s_candidates[valid], e_candidates[valid]):
```

Comment:
can we rename the looping variables `start` and `end`? Pylint is complaining since these variables are already defined in the outer scope.